File: packages/dart/lib/src/utils/text_utils.dart
/// Text utilities for catalogmx
///
/// Provides functions for text normalization, accent removal, and string manipulation
/// used throughout the validators and catalogs.
library;

import 'package:diacritic/diacritic.dart';

/// Removes accents and converts text to uppercase
String cleanText(String text) {
  return removeDiacritics(text).toUpperCase();
}

/// Normalizes text for comparison (removes accents, lowercase, trim)
String normalizeText(String text) {
  return removeDiacritics(text.toLowerCase().trim());
}

/// Removes excluded words from a name
String removeExcludedWords(String text, List<String> excludedWords) {
  final words = text.split(' ');
  final filtered = words.where((word) => !excludedWords.contains(word.toUpperCase())).toList();
  return filtered.join(' ');
}

/// Checks if a string contains only allowed characters
bool containsOnlyAllowed(String text, List<String> allowedChars) {
  for (var i = 0; i < text.length; i++) {
    if (!allowedChars.contains(text[i])) {
      return false;
    }
  }
  return true;
}

/// Gets the first vowel in a string
String? getFirstVowel(String text, {int startIndex = 0}) {
  const vowels = 'AEIOU';
  for (var i = startIndex; i < text.length; i++) {
    if (vowels.contains(text[i])) {
      return text[i];
    }
  }
  return null;
}

/// Gets the first consonant in a string
String? getFirstConsonant(String text, {int startIndex = 0}) {
  const consonants = 'BCDFGHJKLMNPQRSTVWXYZ';
  for (var i = startIndex; i < text.length; i++) {
    if (consonants.contains(text[i])) {
      return text[i];
    }
  }
  return null;
}

/// Cleans a name by removing accents and excluded words
String cleanName(String name, List<String> excludedWords, List<String> allowedChars) {
  // Remove excluded words
  final filtered = removeExcludedWords(name, excludedWords);

  // Clean text and remove accents
  final cleaned = cleanText(filtered);

  // Keep only allowed characters
  final result = StringBuffer();
  for (var i = 0; i < cleaned.length; i++) {
    final char = cleaned[i];
    if (allowedChars.contains(char)) {
      result.write(char);
    } else {
      // Try to remove diacritics
      final normalized = removeDiacritics(char);
      if (allowedChars.contains(normalized)) {
        result.write(normalized);
      } else if (char == ' ') {
        result.write(' ');
      }
    }
  }

  return result.toString().trim();
}

File: packages/kotlin/src/main/kotlin/com/openbancor/catalogmx/utils/TextUtils.kt
package com.openbancor.catalogmx.utils

import com.ibm.icu.text.Normalizer2

/**
 * Text utility functions for Mexican document processing
 */
object TextUtils {
    private val normalizer = Normalizer2.getNFDInstance()

    /**
     * Removes diacritics from text (e.g., á -> a, ñ -> n)
     * Preserves Ñ when specified
     */
    @JvmStatic
    @JvmOverloads
    fun removeDiacritics(text: String, preserveN: Boolean = false): String {
        val normalized = normalizer.normalize(text)
        val result = StringBuilder()
        for (char in normalized) {
            when {
                char == 'Ñ' || char == 'ñ' -> {
                    if (preserveN) result.append(char)
                    else result.append(if (char == 'Ñ') 'N' else 'n')
                }
                Character.getType(char) != Character.NON_SPACING_MARK.toInt() -> {
                    result.append(char)
                }
            }
        }
        return result.toString()
    }

    /**
     * Normalizes text for comparison (uppercase, no accents, trim)
     */
    @JvmStatic
    fun normalizeText(text: String): String =
        removeDiacritics(text.uppercase().trim())

    /**
     * Normalizes text for search (lowercase, no accents, trim)
     * Alias for search operations that need case-insensitive matching
     */
    @JvmStatic
    fun normalize(text: String): String =
        removeDiacritics(text.lowercase().trim())

    /**
     * Cleans a name by removing excluded words and keeping only allowed characters
     */
    @JvmStatic
    fun cleanName(
        name: String,
        excludedWords: List<String>,
        allowedChars: List<Char>
    ): String {
        var cleanedName = removeDiacritics(name.uppercase().trim(), preserveN = true)

        // Split into words
        val words = cleanedName.split(Regex("\\s+")).filter { it.isNotEmpty() }

        // Filter out excluded words
        val filteredWords = words.filter { word ->
            word !in excludedWords
        }

        // Rejoin words
        cleanedName = if (filteredWords.isEmpty()) words.joinToString(" ") else filteredWords.joinToString(" ")

        // Keep only allowed characters
        val result = StringBuilder()
        for (char in cleanedName) {
            if (char in allowedChars || char == ' ') {
                result.append(char)
            }
        }

        return result.toString().trim()
    }

    /**
     * Gets the first vowel from a string starting at a given index
     */
    @JvmStatic
    @JvmOverloads
    fun getFirstVowel(text: String, startIndex: Int = 0): Char? {
        val vowels = "AEIOU"
        for (i in startIndex until text.length) {
            if (text[i].uppercaseChar() in vowels) {
                return text[i].uppercaseChar()
            }
        }
        return null
    }

    /**
     * Gets the first consonant from a string starting at a given index
     */
    @JvmStatic
    @JvmOverloads
    fun getFirstConsonant(text: String, startIndex: Int = 0): Char? {
        val consonants = "BCDFGHJKLMNPQRSTVWXYZ"
        for (i in startIndex until text.length) {
            if (text[i].uppercaseChar() in consonants) {
                return text[i].uppercaseChar()
            }
        }
        return null
    }
}
